lib/Sema/CSApply.cpp

```diff
@@ -2400,10 +2400,37 @@ namespace {
                                          cs.getConstraintLocator(base));
         }
 
+        Type toType = simplifyType(expr->getType());
+        
+        // Don't allow lvalues when indexing into a tuple expr.
+        // If we get any lvalues, add load exprs to convert the tuple to be all rvalues.
+        if (auto *tupleExpr = dyn_cast<TupleExpr>(base)) {
+          unsigned count = tupleExpr->getNumElements();
+          unsigned lvalues = 0;
+          auto &tc = cs.getTypeChecker();
+          SmallVector<TupleTypeElt, 4> tupleElts;
+          for (unsigned i = 0; i < count; i++) {
+            Expr *elementExpr = tupleExpr->getElement(i);
+            Type elementType = elementExpr->getType();
+            if (elementType->isLValueType()) {
+              lvalues++;
+              elementExpr->propagateLValueAccessKind(AccessKind::Read, true);
+              elementExpr = new (tc.Context) LoadExpr(elementExpr, elementType->getRValueType());
+              tupleExpr->setElement(i, elementExpr);
+            }
+            tupleElts.push_back(elementExpr->getType());
+          }
+          
+          if (lvalues > 0) {
+            auto &Context = tupleExpr->getType()->getASTContext();
+            tupleExpr->setType(TupleType::get(tupleElts, Context));
+            toType = toType->getRValueType();
+          }
+        }
+
         return new (cs.getASTContext()) TupleElementExpr(base, dotLoc,
```

Comment:
Would it make sense to allow TupleElementExpr to propagate lvalue-ness, and load the result of that instead? Otherwise, you might end up loading lvalues unnecessarily.

Or would this only come up in the useless cases you mentioned like ```(x, y).1 = z```?